
# Load R packages
```{r}
library(tidyverse)
library(RColorBrewer)
library(adegenet)
library(mclust)
library(MASS)
library(ggforce)
library(cluster)
library(factoextra)
library(gridExtra)
library(randomForest)
library(PCDimension)
library(tsne)
library(vcfR)
library(LEA)
library(tess3r)
```

# Process vcf file
```{r}
## Load specimen data and VCF
ind_data <- read_table("/home/tcm0036/distichus-ddRAD/info/distichus-popmap-master.tsv", col_names = TRUE)

vcfR <- read.vcfR("/mmfs1/scratch/tcm0036/distichus-ddRAD/analyses/population-structure/VAE/distichus-brevirostris-nomissing_pruned.vcf")
    dim(vcfR@gt)

##  Filter out animals excluded in vcfR object from population map 
vcfR_samples <- colnames(vcfR@gt)
vcfR_samples <- gsub("(_/).*", "\\1", vcfR_samples) %>% str_replace("_/", "")
# a bit hacky, but this gets rid of the duplicated path names serving
# as individual IDs in the `vcfR` object, which is a byproduct of plink

## exclude specimens dropped during filtering:
## 266166-8 (altavelensis), 4403-4 (ravitergum), 6081 (dominicensis 4), 866 (ignigularis)
ind_data <- ind_data[ind_data$File_path %in% vcfR_samples,]
    dim(ind_data)

ind_data2 <- ind_data[ind_data$Island == "Hispaniola", ]

write_delim(x = ind_data,
    file = "/home/tcm0036/distichus-ddRAD/info/distichus-popmap-cleaned-master.tsv", # UPDATE PATH
    col_names = TRUE)

write_delim(x = ind_data2,
    file = "/home/tcm0036/distichus-ddRAD/info/all-distichoids-Hispaniola-only-cleaned-popmap.tsv", # UPDATE PATH
    delim = "\t",
    col_names = TRUE)

    # filter out missing SNPs
    vcfR@fix<-vcfR@fix[rowSums(is.na(vcfR@gt)) == 0,]
    vcfR@gt<-vcfR@gt[rowSums(is.na(vcfR@gt)) == 0,]
        dim(vcfR) # note: the number of SNPs after running these commands was reduced from the VCF lacking missing data generated with vcftools

    ## output filtered VCF for other analyses
    write.vcf(x = vcfR, file = "../distichus-spgrp-no-missing-LD-pruned.vcf.gz")

    # convert vcfR object to genind and genlight objects
    data <- vcfR2genind(vcfR)
    gen <- vcfR2genlight(vcfR)

    # scale dataframe
    data_scaled <- scaleGen(data, center=FALSE, scale=FALSE, NA.method=c("mean"), nf)
    data_scaled <- scaleGen(data, center=FALSE, scale=FALSE)

    gen@ind.names <- ind_data$Sample_ID_pop     # First, replace the individual names in the genlight object with those from the specimen information dataframe
    pop(gen) <- ind_data$PopID # Next add population identifiers
```

# Use PCA to visualize population structure
```{r}
################################################################
## Perform Principal Component Analysis

    # This performs PCA on the genlight object `gen` and requires manually choosing the number of principal components to retain
    gen_pca <- glPca(gen) # I chose 6
    gen_pca_scores <- as_tibble(as.data.frame(gen_pca$scores))
    gen_pca_scores$pop <- gen@pop
    gen_pca_scores$ind <- gen@ind.names

    ## Set color palette
    colorCount <- length(unique(gen@pop))
    myColors <- colorRampPalette(brewer.pal(11, "RdYlBu"))(colorCount) # could be more distinctive still
    
    p <- ggplot(gen_pca_scores, aes(x = PC1, y = PC2, color = pop))
        p <- p + geom_point(size = 2) + scale_colour_manual(values = myColors)
        #p <- p + stat_ellipse(level = 0.95, size = 1)
        p <- p + geom_hline(yintercept = 0)
        p <- p + geom_vline(xintercept = 0)
        p <- p + theme_bw()
    p
```

# Run discriminant analysis of principal components to identify the optimal value of K
```{r}
################################################################
## Perform DAPC (discriminant analysis of principal components)

    set.seed(15)
    # # This will prompt two questions to determine the number of PCs to retain (350) 
    # # and the value of K to keep based on BIC scores
    # grp <- find.clusters(gen, max.n.clust = 30)
    # dapc1 <- dapc(gen, grp$grp, n.pca = 150, n.da = 5)
    # # save individual cluster assignment as file
    # write.table(dapc1$grp, "DAPC-group-assignments.txt", quote = F, sep = "\t", col.names = F)

    # # plot clusters for first two LDs
    # dapc.plot.df <- as.data.frame(dapc1$ind.coord)
    # write.table(dapc.plot.df, "dapc-K10-5LDs.txt", quote=F, sep="\t", col.names=TRUE) # need column header for sample name column
    grp <- find.clusters(gen, max.n.clust = 20, n.pca = 300, choose.n.clust = FALSE, criterion = "diffNgroup", stat = "BIC")
    dapc.xval <- xvalDapc(data_scaled, grp$grp, training.set = 0.9)
    xval.plot.df <- as.data.frame(dapc.xval$DAPC$ind.coord)
    #rownames(xval.plot.df) <- ind_data$Sample_ID_pop
    write_delim(x = cbind(ind_data$Sample_ID_pop, xval.plot.df),
                    file = "xval_DAPC_K8.tsv",
                    delim = "\t", col_names = TRUE)
    
     # plot clusters for LDs 1 and 2
    ggplot(data = xval.plot.df, aes(x = LD1, y = LD2, color = grp$grp)) +
    geom_point(cex = 3) +
    theme_classic() +
    labs(x = "Linear discriminant 1", y = "Linear discriminant 2")
    ggsave("LD1-LD2.pdf")

    # plot clusters for next two LDs
    ggplot(data= xval.plot.df, aes(x = LD3, y = LD4, color = grp$grp)) +
    geom_point(cex = 3) +
    theme_classic() +
    labs(x = "Linear discriminant 3", y = "Linear discriminant 4")
    ggsave("LD3-LD4.pdf")
```


# Take cleaned VCF from vcfR and change sample names to be more informative
```{bash}
conda activate genomics_env

# Rename samples in VCF
awk '{print $2}' distichus-popmap-cleaned-master.tsv | tail -n276 > new-sample_names.txt  
bcftools view distichus-spgrp-no-missing-LD-pruned.vcf.gz | bcftools reheader --samples new-sample_names.txt -o distichus-spgrp-no-missing-LD-pruned-informative-names.vcf
```

# Use sNMF (sparse nonnegative matrix factorization) to identify the optimal value of K
```{r}
################################################################
## Infer ancestry proportions with sNMF

setwd("/scratch/tcm0036/distichus-ddRAD/analyses/population-structure/lea")

vcf <- "../distichus-spgrp-no-missing-LD-pruned-informative-names.vcf"
vcf2geno(vcf, output.file = "distichus-spgrp-no-missing-LD-pruned.geno")
#geno <- read.geno("distichus-spgrp-no-missing-LD-pruned.geno")
#   dim(geno)

obj.snmf <- snmf("distichus-spgrp-no-missing-LD-pruned.geno", K = 1:10, project = "new",
    repetitions = 50, tolerance = 0.00001, seed = 16,
    alpha = 100, entropy = TRUE, ploidy = 2)

pdf("snmf_K1-10_50-replicates.pdf")
plot(obj.snmf, cex = 1.2, col = "lightblue", pch = 19)
dev.off()

for (i in 1:10){
    print(mean(cross.entropy(obj.snmf, K = i)))
}
```

# For optimal value of K, plot ancestry coefficients
```{r}

k <- 10
ce <- cross.entropy(obj.snmf, K = k) # K = 10 determined optimal
best_run <- which.min(ce)

qmatrix <- LEA::Q(obj.snmf, run = best_run, K = k)
Qmatrix  <- tess3r::as.qmatrix(qmatrix)

# get cluster assignments per individual
cluster <- apply(qmatrix, 1, which.max)

sNMF_ancestry_coefficients_cluster <- as.data.frame(cbind(ind_data$Sample_ID_pop, cluster, Qmatrix))
write_delim(x = sNMF_ancestry_coefficients_cluster, file = "sNMF_K10_ancestry_coefficients_cluster.tsv",
            col_names = TRUE, delim = "\t")

mypalette <-  tess3r::CreatePalette(color.vector = c("tomato", "chartreuse", "gold", "blue", "gray80",
    "violet", "wheat", "olivedrab", "lightblue", "orange"), palette.length = k)

pdf(paste0("sNMF-K", k, "-barplot.pdf"))
bp <- barplot(Qmatrix, border = NA, space = 0, xlab = "Specimens",
        ylab = "Ancestry",
        col.palette = mypalette, sort.by.Q = T)

#axis(1, at = 1:nrow(Qmatrix), labels = bp$order, las = 3, cex.axis = .5)
axis(1, at = 1:nrow(Qmatrix), labels = bp$order, las = 3, cex.axis = .5)
dev.off()

# Below is working code for a STRUCTURE-type barplot using base R functions applied to the qmatrix for K=9
# snmf <- read_delim("population-structure/snmf/sNMF_K9_ancestry_coefficients_cluster.tsv", delim = "\t")
# test <- snmf[order(snmf$cluster),]
# pdf("population-structure/snmf/sNMF-K9-barplot.pdf")
# barplot(t(test[,3:ncol(test)]), col = c("gray80", "#634102", "#c6b89c", "#04e779", "#fddc1f", "#993cc8", "#cf5d34", "black", "#1a1ac2"), names.arg=test$V1...1, las =2, cex.names = 0.4, border = NA)
# dev.off()

```

####################################################################################
# Process VCF without missing data into STRUCTURE formatted file for conStruct input
```{bash}

cd conStruct/

# Drop samples not from Hispaniola
awk '{print $2}' all-distichoids-Hispaniola-only-cleaned-popmap.tsv | tail -n251 > specimens-to-keep.txt 
vcftools --vcf distichus-spgrp-no-missing-LD-pruned-informative-names.vcf --keep specimens-to-keep.txt --recode --out distichus-spgrp-Hispaniola-no-missing-LD-pruned

# convert VCF to STRUCTURE format
plink --vcf distichus-spgrp-Hispaniola-no-missing-LD-pruned.recode.vcf --double-id --allow-extra-chr --recode structure --out distichus-spgroup-conStruct

# And get rid of two header lines to leave just sample information
tail -n251 distichus-spgroup-conStruct.recode.strct_in > distichus-spgroup-noheader-conStruct.str
```

# Below code is used to run conStruct
```{r}
setwd("/mmfs1/scratch/tcm0036/distichus-ddRAD/analyses/population-structure/conStruct")

library(tidyverse)
library(fields)
library(conStruct)
library(doParallel)
library(foreach)

#########
# Generate input files for conStruct
#########

# (1) Load allele frequencies file
    #structure2conStruct(infile = "distichus-spgroup-noheader-conStruct.str",
    #    onerowperind = TRUE,
    #    start.loci = 3,
    # This data matrix lacks any missing data so I just chose 0
    #    missing.datum = 0,
    #    outfile = "distichus-spgroup-conStuct-input")
load("distichus-spgroup-conStuct-input.RData")

# (2) Load coordinates for specimen sampling localities
ind_data <- read_table("/home/tcm0036/distichus-ddRAD/info/all-distichoids-Hispaniola-only-cleaned-popmap.tsv", col_names = TRUE)
latlong <- as.matrix(ind_data[,c('Longitude', 'Latitude')])

# (3) Generate geographic distance matrix from specimen sampling localities

## As per https://cran.r-project.org/web/packages/conStruct/vignettes/format-data.html#geographic-distance-matrix
## I will be using the `fields::rdist.earth` function to estimate pairwise great-circle distances
## between sampling coordinates
geoDist <- fields::rdist.earth(x1 = latlong, x2 = latlong)

#########
# run conStuct analyses for Ks 1-10
#########

dir.create("/mmfs1/scratch/tcm0036/distichus-ddRAD/analyses/population-structure/conStruct/outputs2")

#cl <- makeCluster(10, type = "FORK")
#registerDoParallel(cl)

myRun <- x.validation(train.prop = 0.9, 
    n.reps = 10,
    K = 1:10,
    freqs = freqs,
    coords = latlong,
    geoDist = geoDist,
    n.iter = 1e4,
    make.figs = TRUE,
    save.files = TRUE,
    parallel = TRUE,
    n.nodes = 10, 
    prefix = "outputs2/distichus-spgroup-CV")

#stopCluster(cl)
```