######################################################################################
# This script includes code for estimating pairwise Fst in the 
# R package `hierfstat`, as well as estimating geographic distance 
# and environmental dissimilarity to quantify the contributions of
# isolation-by-distance and isolation-by-environment to genetic differentiation
# using multivariate statistical approaches.
# Authors: Tanner C. Myers, Pietro L. H. de Mello, Paul M. Hime, and Richard E. Glor
######################################################################################

```{r}
# Change below to reflect working directory
# work_dir <- getwd()
# setwd(work_dir)
setwd("/scratch/tcm0036/distichus-ddRAD/analyses/environmental-association")

rm(list = ls())

# Load R packages
library(tidyverse)
library(reshape2)
library(vcfR)
library(adegenet)
library(hierfstat)
library(sp)
library(raster)
library(rgeos)
library(rgdal)
library(adespatial)
library(envirem)
library(RStoolbox)
library(BEDASSLE)
library(ecodist) # MRM function may be able to replace MMRR
library(gdm)
library(vegan)
library(lavaan)
#library(devtools)
#devtools::install_github("cran/simba")
library(simba)

# Load functions
## Load MMRR function available from https://datadryad.org/stash/dataset/doi:10.5061%2Fdryad.kt71r
source("~/distichus-ddRAD/scripts/MMRR.R")

## Load quickMem function from Borcard et al. (2018)
## https://raw.githubusercontent.com/zdealveindy/anadat-r/master/scripts/NumEcolR2/quickMEM.R
source("~/distichus-ddRAD/scripts/quickMEM.R")

## Load function for imputing NAs in the raster from
## https://stackoverflow.com/questions/45641168/fill-in-gaps-e-g-not-single-cells-of-na-values-in-raster-using-a-neighborhood
fill.na <- function(x) {
  center = 0.5 + (width*width/2)
  if(is.na(x)[center]) {
    return(mean(x, na.rm = TRUE))
  } else {
    return(x[center])
  }
}
```

############################################################
# Load datasets
```{r}
# Load genetic and morphological datasets and sampling info

# Load ddRADseq individual information
RAD_data <- read_table("/home/tcm0036/distichus-ddRAD/info/distichus-popmap-cleaned-master.tsv", col_names = TRUE)
## Subset dataset to include only Hispaniolan members of the distichus species group
RAD_data2 <- RAD_data %>% filter(Sample != "4481" & Sample != "10085" & Sample != "10086") %>%
    filter(!(Taxon %in% c("websteri", "marron", "caudalis", "altavelensis"))) %>%
    filter(!grepl("_rep", Sample)) %>% # Temporary, eventually I need to deal with the technical replicates in the dataset
    filter(Island %in% "Hispaniola")

snmfK9 <- read_table("/scratch/tcm0036/distichus-ddRAD/analyses/population-structure/lea/sNMF_K9_ancestry_coefficients_cluster-specimen-cleaned.tsv")
    names(snmfK9)[names(snmfK9) == "V1"] <- "Sample_ID_pop"

RAD_data2 <- inner_join(snmfK9, RAD_data2)

# Load linkage-pruned VCF file with 0% missing data
vcfR <- read.vcfR("../population-structure/distichus-spgrp-no-missing-LD-pruned-informative-names.vcf")

# convert vcfR object to adegenet genind object
    dat <- vcfR2genind(vcfR)
    ## Subset genind object to reflect individuals retained in subsetted information file
    dat2 <- dat[indNames(dat) %in% RAD_data2$Sample_ID_pop]
    ## Set population to be sampling locality
    pop(dat2) <- RAD_data2$Locality
```

#######################################################################
# Estimate genetic distance (i.e., pairwise Fst (Weir & Cockerham 1984))
```{r}
# Estimate pairwise Fst for genetic differentiation matrix

data <- genind2hierfstat(dat = dat2, pop = dat2@pop)
## Add population column
data <- as.data.frame(cbind(dat2@pop, data[, c(2:ncol(data))]))

gen_diff <- pairwise.WCfst(dat = data, diploid = TRUE)
    save(gen_diff, file = "Fst-matrix.RData")
    load("mmrr/Fst-matrix.RData")
    ## Changes NAs to 0s
    gen_diff[is.na(gen_diff)] = 0

gen_dist <- gen_diff/(1 - gen_diff)
    class(gen_dist)
    isSymmetric(gen_dist) # Needs to be TRUE
```

############################################################
# Estimate geographic distance
```{r}
# Infer Euclidean distance matrix from sampling points
## Subset unique locations
coords <- distinct(RAD_data2[, c("Longitude", "Latitude")])
coords <- as.matrix(coords)

geog_dist <- sp::spDists(x = coords, y = coords, longlat = TRUE)
```

##############################################################
# Prepare environmental variables
```{r}
# Load environmental dataset

## Load shapefile
DOM <- getData('GADM', country = 'DOM', level=0, path = paste0("~/distichus-spatial-variation/data/shape-files/"), download = F)
HTI <- getData('GADM', country = 'HTI', level=0, path = paste0("~/distichus-spatial-variation/data/shape-files/"), download =F)
row.names(DOM) <- paste("DOM", row.names(DOM), sep = "_")
row.names(HTI) <- paste("HTI", row.names(HTI), sep = "_")
Hispaniola <- rbind(HTI, DOM, makeUniqueIDs = TRUE)
Hispaniola <- gSimplify(Hispaniola, tol = 0.01, topologyPreserve = TRUE)

## Load environmental variables
chelsa_clim <- raster::stack(list.files(path = "~/distichus-spatial-variation/data/chelsa_new/", pattern = ".asc", full.names = TRUE))
modis_vi <- raster::stack(list.files(path = "~/distichus-spatial-variation/data/MODIS_new/", pattern = ".asc", full.names = TRUE))
## Raster stacks have different extents
    ## Since the extent of modis_vi fits within chelsa_clim's extent, crop chelsa_clim
    chelsa_clim <- crop(chelsa_clim, modis_vi@extent)
    chelsa_clim@extent <- raster::alignExtent(chelsa_clim@extent, modis_vi)

## Now, load elevational data that has already been adjusted to the MODIS variable extent
elev_srtm <- raster("~/distichus-spatial-variation/data/elevation_new/SRTM_elevation_1km.asc")

## Merge all environmental data layers into single raster stack
env <- raster::stack(elev_srtm, chelsa_clim, modis_vi, full.names = TRUE)
   env <- raster::crop(env, extent(Hispaniola))

## Check individual variable distributions,
# for (i in 1:20){
#     pdf(paste0("temp_plots/", names(env[[i]]), "-histogram.pdf"))
#         # plot histograms for each raster
#         hist(env[[i]], col = "green")
#     dev.off()
# }
```

############################################################
# Extract values from environmental variables
```{r}
# Extract values for raw environmental variables

## use function above to impute NA values for localities not represented in raster layers
width <- 5

for (i in 1:20){
    var <- terra::focal(env[[i]], w = matrix(1, width, width), fun = fill.na, na.rm = FALSE, pad = TRUE)
        assign(paste0("var_", i), var)
}

# Add imputed rasters to a new raster stack
env_imputed <- stack(var_1, var_2, var_3, var_4, var_5, var_6, var_7, var_8, var_9, var_10, 
                    var_11, var_12, var_13, var_14, var_15, var_16, var_17, var_18, var_19, var_20)
    ## Change layer names to original variable names
    names(env_imputed) <- names(env)

## Extract values
env_vals <- terra::extract(env_imputed, coords)
    env_df <- as_tibble(data.frame(cbind(coords, env_vals)))
    write_delim(x = env_df, file = "mmrr/raw-environmental-variables.csv", delim = ",", col_names = TRUE)
    env_df <- read_csv("mmrr/raw-environmental-variables.csv", col_names = TRUE)
```

###############################################################
# Perform Principal Component Analysis on environment variables
```{r}
# Perform environmental data layer PCA

## Perform raster pca
pca_env <- RStoolbox::rasterPCA(env, spca = TRUE)
    plot(pca_env$map)
    summary(pca_env$model)
    knitr::kable(round(pca_env$model$loadings[, 1:5], 5)) # top 5 loadings

## use function above to impute NA values in each principal component raster
width <- 5

for (i in 1:5){
    pc <- terra::focal(pca_env$map[[i]], w = matrix(1, width, width), fun = fill.na, na.rm = FALSE, pad = TRUE)
        assign(paste0("pc", i), pc)
}

pc1_5 <- stack(pc1, pc2, pc3, pc4, pc5)
    names(pc1_5) <- c("PC1", "PC2", "PC3", "PC4", "PC5")

## Extract values for first 5 principal components
pc_vals <- terra::extract(pc1_5, coords)
    pc_df <- as_tibble(data.frame(cbind(coords, pc_vals)))
    write_delim(x = pc_df, file = "mmrr/rasterPCs.csv", delim = ",", col_names = TRUE)
    pc_df <- read_csv("mmrr/rasterPCs.csv", col_names = TRUE)
```

##############################################################
# Estimate environmental dissimilarity
```{r}
# Estimate Euclidean distances between environmental PCs
pc_dist <- as.matrix(dist(pc_df[, c("PC1", "PC2", "PC3", "PC4", "PC5")]), method = "euclidean")

# Now, do the same thing, but for raw environmental variables
env_dist <- as.matrix(dist(env_df[, 3:ncol(env_df)]))

# Estimate distances between environmental variables that capture enviromental variation across Hispaniola:
# annual mean air temperature, temperature seasonality, annual precipitation amount, and precipitation seasonality
env_sub_dist <- as.matrix(dist(env_df[, c("CHELSA_bio10_01", "CHELSA_bio10_04", "CHELSA_bio10_12", "CHELSA_bio10_15")]))
```

##############################################################
# Multiple Matrix Regression with Randomization
```{r}
# Run MMRR using raw ecological variables

# (1) Quantify contributions of IBD (great circle distance)
# and IBE (Euclidean distances between 20 environmental variables)
## Define distance matrices to be used as dependent variables
xvars <- list(ecology = env_dist, geography = geog_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (2) Quantify contributions of IBD (great circle distance) and IBE (Subsetted environmental variables)
xvars <- list(geography = geog_dist, ecology = env_sub_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (3) Only IBD
xvars <- list(geography = geog_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (4) Only IBE
xvars <- list(ecology = env_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (5) Only IBE (Subsetted environmental variables)
xvars <- list(ecology = env_sub_dist)
MMRR(gen_dist, xvars, nperm = 1000)
```

```{r}
# Run MMRR using principal components for ecological variables

# (1) Quantify contributions of IBD (great circle distance) and IBE (Euclidean distances between 5 environmental PCs in aggregate)
## Define distance matrices to be used as dependent variables
xvars <- list(ecology = pc_dist, geography = geog_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (2) Quantify contributions of IBD (great circle distance) and IBE (PCs disaggregated)
xvars <- list(geography = geog_dist,
            pc1 = as.matrix(dist(pc_df$PC1, method = "euclidean")),
            pc2 = as.matrix(dist(pc_df$PC2, method = "euclidean")),
            pc3 = as.matrix(dist(pc_df$PC3, method = "euclidean")),
            pc4 = as.matrix(dist(pc_df$PC4, method = "euclidean")),
            pc5 = as.matrix(dist(pc_df$PC5, method = "euclidean")))
MMRR(gen_dist, xvars, nperm = 1000)

# (3) Only IBD
xvars <- list(geography = geog_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (4) Only IBE
xvars <- list(ecology = pc_dist)
MMRR(gen_dist, xvars, nperm = 1000)

# (5) Only IBE, with individuals PCs treated as distinct variables
xvars <- list(pc1 = as.matrix(dist(pc_df$PC1, method = "euclidean")),
            pc2 = as.matrix(dist(pc_df$PC2, method = "euclidean")),
            pc3 = as.matrix(dist(pc_df$PC3, method = "euclidean")),
            pc4 = as.matrix(dist(pc_df$PC4, method = "euclidean")),
            pc5 = as.matrix(dist(pc_df$PC5, method = "euclidean")))
MMRR(gen_dist, xvars, nperm = 1000)
```

################################################################
# Redundancy Analysis
```{r}
# First, estimate dbMEMs for use as a covariate in RDA
latlong <- SpatialPoints(RAD_data2[, c("Longitude", "Latitude")], proj4string = CRS("+proj=longlat"))

distichoid_xy <- as.data.frame(spTransform(latlong, CRS("+proj=longlat +datum=WGS84 +ellps=WGS84 +towgs84=0,0,0")))

dbmem.1 <- quickMEM(dat2@tab, distichoid_xy, perm.max = 999)
    load("rda/dbMEM_1.RData")

# RDA

all_data <- inner_join(RAD_data2, env_df)

rda.1 <- rda(dat2@tab ~., data = all_data[, 12:ncol(all_data)], scale = TRUE)
    RsquareAdj(rda.1)
    summary(rda.1)$concont

vif.cca(rda.1)

signif.rda.1 <- anova.cca(rda.1)

levels(all_data$cluster) <- c("ignigularis", "east dominicensis", "South paleo-island", "properus", "brevirostris 1", "west dominicensis", "brevirostris 2", "ravitergum")
clust <- all_data$cluster
bg <- c("#cf5d34", "#634102", "#993cc8", "gray80", "black", "#c6b89c", "gray27", "#fddc1f")

# bg <- c("#634102", "gray80", "#cf5d34", "black", "#634102", "#993cc8",
#         "#fddc1f", "#fddc1f", "#c6b89c", "#1a1ac2", "#c6b89c", "#1a1ac2",
#         "#1a1ac2", "#1a1ac2")

pdf("rda/rda-full-plot.pdf")
    plot(rda.1, type = "n", scaling = 3)
    points(rda.1, display = "species", pch = 20, cex = 0.7, col = "gray32", scaling = 3, choices = c(1, 2)) # SNPs in this case
    #points(rda.1, display = "sites", pch = 21, cex = 1.3, col = "gray32", scaling = 3, bg = bg[as.factor(l)], choices = c(1, 3)) # lizards
    points(rda.1, display = "sites", pch = 21, cex = 1.3, col = "gray32", scaling = 3, bg = bg[as.factor(clust)], choices = c(1, 2)) # lizards
    text(rda.1, scaling = 3, display = "bp", col = "#0868ac", cex = 1) # predictor variables
    legend("topright", legend = levels(clust), bty = "n", col = "gray32", pch = 21, cex = 1, pt.bg = bg)
dev.off()

## Forward selection with environmental variables

fwd.sel <- ordiR2step()

# RDA constrained on space

```

############################################################
# Structural equation modeling (SEM)
```{r}
# SEM

## Create genlight object and subset individuals
gen <- vcfR2genlight(vcfR)
gen2 <- gen[gen@ind.names %in% RAD_data2$Sample_ID_pop]

# Create distance matrices from geographic distance and environmental variables
    ## `melt` is used to transform distance matrices into dataframes and I am
    ## keeping only the third column with the distance estimates
geo.dist <- melt(base::scale(dist(distichoid_xy, method = "euclidean")))[, 3]
env.dist.1 <- melt(base::scale(dist(all_data$CHELSA_bio10_01, method = "euclidean")))[, 3]
env.dist.2 <- melt(base::scale(dist(all_data$CHELSA_bio10_02, method = "euclidean")))[, 3]
env.dist.3 <- melt(base::scale(dist(all_data$CHELSA_bio10_03, method = "euclidean")))[, 3]
env.dist.4 <- melt(base::scale(dist(all_data$CHELSA_bio10_04, method = "euclidean")))[, 3]
env.dist.5 <- melt(base::scale(dist(all_data$CHELSA_bio10_05, method = "euclidean")))[, 3]
env.dist.6 <- melt(base::scale(dist(all_data$CHELSA_bio10_06, method = "euclidean")))[, 3]
env.dist.7 <- melt(base::scale(dist(all_data$CHELSA_bio10_07, method = "euclidean")))[, 3]
env.dist.8 <- melt(base::scale(dist(all_data$CHELSA_bio10_10, method = "euclidean")))[, 3]
env.dist.9 <- melt(base::scale(dist(all_data$CHELSA_bio10_11, method = "euclidean")))[, 3]
env.dist.10 <- melt(base::scale(dist(all_data$CHELSA_bio10_12, method = "euclidean")))[, 3]
env.dist.11 <- melt(base::scale(dist(all_data$CHELSA_bio10_13, method = "euclidean")))[, 3]
env.dist.12 <- melt(base::scale(dist(all_data$CHELSA_bio10_14, method = "euclidean")))[, 3]
env.dist.13 <- melt(base::scale(dist(all_data$CHELSA_bio10_15, method = "euclidean")))[, 3]
env.dist.14 <- melt(base::scale(dist(all_data$CHELSA_bio10_16, method = "euclidean")))[, 3]
env.dist.15 <- melt(base::scale(dist(all_data$CHELSA_bio10_17, method = "euclidean")))[, 3]
env.dist.16 <- melt(base::scale(dist(all_data$SRTM_elevation_1km, method = "euclidean")))[, 3]
env.dist.17 <- melt(base::scale(dist(all_data$march_EVI_mean, method = "euclidean")))[, 3]
env.dist.18 <- melt(base::scale(dist(all_data$march_NDVI_mean, method = "euclidean")))[, 3]
env.dist.19 <- melt(base::scale(dist(all_data$may_EVI_mean, method = "euclidean")))[, 3]
env.dist.20 <- melt(base::scale(dist(all_data$may_NDVI_mean, method = "euclidean")))[, 3]

# Create distance matrix from genetic data
gen.dist <- melt(base::scale(dist(gen2, method = "euclidean")))
    colnames(gen.dist) <- c("Specimen1", "Specimen2", "gen.dist")

# Now, merge all distance matrices into a single dataframe
full <- data.frame(gen.dist, env.dist.1, env.dist.2, env.dist.3, env.dist.4, env.dist.5, env.dist.6,
                env.dist.7, env.dist.8, env.dist.9, env.dist.10, env.dist.11, env.dist.12, env.dist.13,
                env.dist.14, env.dist.15, env.dist.16, env.dist.17, env.dist.18, env.dist.19, env.dist.20,
                geo.dist)

## Define model framework for SEM
    ## =~ operator defines latent variables
    ## ~~ variance/covariance operator
    ## This model specifies genetic distance as as the dependent variable
    ## of two latent variables: environmental dissimilarity and geographic
    ## distance
model_all <- '
    env =~ env.dist.1+env.dist.2+env.dist.3+env.dist.4+env.dist.5+env.dist.6+env.dist.7+env.dist.8+env.dist.9+env.dist.10+env.dist.11+env.dist.12+env.dist.13+env.dist.14+env.dist.15+env.dist.16+env.dist.17+env.dist.18+env.dist.19+env.dist.20
    geo =~ geo.dist

    gen.dist ~ env + geo
    geo ~~ env
    env ~~ env
    geo ~~ geo
'

fit_all <-  sem(model_all, data = full, estimator = "MLR", se = "robust.huber.white")
s <- summary(fit_all, standardized = TRUE, fit.measures = TRUE, rsquare = TRUE)
p <- parameterEstimates(fit_all)
```

```{r}
# Perform generalized dissimilarity modeling

# pop <- row.names(gen_dist)
# gen_gdm <- as.data.frame(cbind(pop, gen_dist))

# env_gdm <- as_tibble(cbind(pop, coords, pc_dist))

# gdm_data <- formatsitepair(bioData = gen_gdm, bioFormat = 3,
#                         predData = env_gdm, siteColumn = "pop",
#                         XColumn = "Longitude", YColumn = "Latitude")

# # Run GDM with variable selection
# gdm.varImp(gdm_data, geo = TRUE, nPerm = 100)
## Received following error:
# Error in gdm.varImp(gdm_data, geo = TRUE, nPerm = 100) : 
#   The spTable contains negative distance values. Must be between 0 - 1.
```

###################################################################################################
# The code above this point analyzes data involving both Anolis distichus subspecies and
# A. brevirostris. The code below is focused on a subset of the data including only the subspecies
# of A. distichus found on Hispaniola: A. d. aurifer, A. d. dominicensis, A. d. favillarum, 
# A. d. ignigularis, A. d. properus, A. d. ravitergum, A. d. suppar, and A. d. vinosus.
###################################################################################################

```{r}
# Subset datasets above for distichus subspecies only

## Exclude individual specimens that cluster with A. brevirostris
RAD_data3 <- RAD_data2[RAD_data2$Sample_ID_pop %in% snmfK9$V1[snmfK9$cluster != 6 & snmfK9$cluster != 8], ]
RAD_data3 <- RAD_data2[RAD_data2$cluster != 6 & RAD_data2$cluster != 6, ]

## Subset genind object to reflect individuals retained in subsetted information file
    dat3 <- dat[indNames(dat) %in% RAD_data3$Sample_ID_pop]
    ## Set population to be sampling locality
    pop(dat3) <- RAD_data3$Locality

# A perk of this dataset is that it allows incorporation of the 
# morphometric dataset from Myers et al. (2020). 
# Spreadsheet with measurements available from: https://doi.org/10.5061/dryad.qfttdz0dq
morph_data <- read_csv("mmrr/Myers_2020_distichus_morphdata.csv", col_names = TRUE)
    ## Subset by traits
    morph_data <- morph_data[, c("Latitude", "Longitude", "SVL")]
```
```{r}
# Re-estimate distance matrices

## Geographic distance
coords2 <- distinct(RAD_data3[, c("Longitude", "Latitude")])
coords2 <- as.matrix(coords2)

geog_dist2 <- sp::spDists(x = coords2, y = coords2, longlat = TRUE)

## Environmental distance
env_df2 <- inner_join(as_tibble(coords2), env_df)

# Now, do the same thing, but for raw environmental variables
env_dist2 <- as.matrix(dist(env_df2[, 3:ncol(env_df2)]))

# Estimate distances between environmental variables that capture enviromental variation across Hispaniola:
# annual mean air temperature, temperature seasonality, annual precipitation amount, and precipitation seasonality
env_sub_dist2 <- as.matrix(dist(env_df2[, c("CHELSA_bio10_01", "CHELSA_bio10_04", "CHELSA_bio10_12", "CHELSA_bio10_15")]))
## Genetic distance

# Estimate pairwise Fst for genetic differentiation matrix

data <- genind2hierfstat(dat = dat3, pop = dat3@pop)
data <- as.data.frame(cbind(dat3@pop, data[, c(2:ncol(data))]))

gen_diff2 <- pairwise.WCfst(dat = data, diploid = TRUE)
    save(gen_diff, file = "mmrr/Fst-matrix-distichus-subsp.RData")
    load("mmrr/Fst-matrix-distichus-subsp.RData")
    ## Changes NAs to 0s
    gen_diff2[is.na(gen_diff2)] = 0

gen_dist2 <- gen_diff2/(1 - gen_diff2)
    class(gen_dist2)
    isSymmetric(gen_dist2) # Needs to be TRUE

## Morphological distance

```

##############################################################
# Multiple Matrix Regression with Randomization
```{r}
# Run MMRR using raw ecological variables

# (1) Quantify contributions of IBD (great circle distance)
# and IBE (Euclidean distances between 20 environmental variables)
## Define distance matrices to be used as dependent variables
xvars <- list(ecology = env_dist2, geography = geog_dist2)
MMRR(gen_dist2, xvars, nperm = 1000)

# (2) Quantify contributions of IBD (great circle distance) and IBE (Subsetted environmental variables)
xvars <- list(geography = geog_dist2, ecology = env_sub_dist2)
MMRR(gen_dist2, xvars, nperm = 1000)

# (3) Only IBD
xvars <- list(geography = geog_dist2)
MMRR(gen_dist2, xvars, nperm = 1000)

# (4) Only IBE
xvars <- list(ecology = env_dist2)
MMRR(gen_dist2, xvars, nperm = 1000)

# (5) Only IBE (Subsetted environmental variables)
xvars <- list(ecology = env_sub_dist2)
MMRR(gen_dist2, xvars, nperm = 1000)
```


############################################################
# Structural equation modeling (SEM)
```{r}
# SEM

all_data2 <- inner_join(RAD_data3, env_df2)

## Create genlight object and subset individuals
gen3 <- gen[gen@ind.names %in% RAD_data3$Sample_ID_pop]

    ## Set population to be sampling locality
    pop(dat3) <- RAD_data3$Locality

latlong <- SpatialPoints(RAD_data3[, c("Longitude", "Latitude")], proj4string = CRS("+proj=longlat"))

distichus_xy <- as.data.frame(spTransform(latlong, CRS("+proj=longlat +datum=WGS84 +ellps=WGS84 +towgs84=0,0,0")))

# Create distance matrices from geographic distance and environmental variables
    ## `melt` is used to transform distance matrices into dataframes and I am
    ## keeping only the third column with the distance estimates
geo.dist <- melt(base::scale(dist(distichus_xy, method = "euclidean")))[, 3]
env.dist.1 <- melt(base::scale(dist(all_data2$CHELSA_bio10_01, method = "euclidean")))[, 3]
env.dist.2 <- melt(base::scale(dist(all_data2$CHELSA_bio10_02, method = "euclidean")))[, 3]
env.dist.3 <- melt(base::scale(dist(all_data2$CHELSA_bio10_03, method = "euclidean")))[, 3]
env.dist.4 <- melt(base::scale(dist(all_data2$CHELSA_bio10_04, method = "euclidean")))[, 3]
env.dist.5 <- melt(base::scale(dist(all_data2$CHELSA_bio10_05, method = "euclidean")))[, 3]
env.dist.6 <- melt(base::scale(dist(all_data2$CHELSA_bio10_06, method = "euclidean")))[, 3]
env.dist.7 <- melt(base::scale(dist(all_data2$CHELSA_bio10_07, method = "euclidean")))[, 3]
env.dist.8 <- melt(base::scale(dist(all_data2$CHELSA_bio10_10, method = "euclidean")))[, 3]
env.dist.9 <- melt(base::scale(dist(all_data2$CHELSA_bio10_11, method = "euclidean")))[, 3]
env.dist.10 <- melt(base::scale(dist(all_data2$CHELSA_bio10_12, method = "euclidean")))[, 3]
env.dist.11 <- melt(base::scale(dist(all_data2$CHELSA_bio10_13, method = "euclidean")))[, 3]
env.dist.12 <- melt(base::scale(dist(all_data2$CHELSA_bio10_14, method = "euclidean")))[, 3]
env.dist.13 <- melt(base::scale(dist(all_data2$CHELSA_bio10_15, method = "euclidean")))[, 3]
env.dist.14 <- melt(base::scale(dist(all_data2$CHELSA_bio10_16, method = "euclidean")))[, 3]
env.dist.15 <- melt(base::scale(dist(all_data2$CHELSA_bio10_17, method = "euclidean")))[, 3]
env.dist.16 <- melt(base::scale(dist(all_data2$SRTM_elevation_1km, method = "euclidean")))[, 3]
env.dist.17 <- melt(base::scale(dist(all_data2$march_EVI_mean, method = "euclidean")))[, 3]
env.dist.18 <- melt(base::scale(dist(all_data2$march_NDVI_mean, method = "euclidean")))[, 3]
env.dist.19 <- melt(base::scale(dist(all_data2$may_EVI_mean, method = "euclidean")))[, 3]
env.dist.20 <- melt(base::scale(dist(all_data2$may_NDVI_mean, method = "euclidean")))[, 3]

# Create distance matrix from genetic data
gen.dist <- melt(base::scale(dist(gen3, method = "euclidean")))
    colnames(gen.dist) <- c("Specimen1", "Specimen2", "gen.dist")

# Now, merge all distance matrices into a single dataframe
full <- data.frame(gen.dist, env.dist.1, env.dist.2, env.dist.3, env.dist.4, env.dist.5, env.dist.6,
                env.dist.7, env.dist.8, env.dist.9, env.dist.10, env.dist.11, env.dist.12, env.dist.13,
                env.dist.14, env.dist.15, env.dist.16, env.dist.17, env.dist.18, env.dist.19, env.dist.20,
                geo.dist)

## Define model framework for SEM
    ## =~ operator defines latent variables
    ## ~~ variance/covariance operator
    ## This model specifies genetic distance as as the dependent variable
    ## of two latent variables: environmental dissimilarity and geographic
    ## distance
model_all <- '
    env =~ env.dist.1+env.dist.2+env.dist.3+env.dist.4+env.dist.5+env.dist.6+env.dist.7+env.dist.8+env.dist.9+env.dist.10+env.dist.11+env.dist.12+env.dist.13+env.dist.14+env.dist.15+env.dist.16+env.dist.17+env.dist.18+env.dist.19+env.dist.20
    geo =~ geo.dist

    gen.dist ~ env + geo
    geo ~~ env
    env ~~ env
    geo ~~ geo
'

fit_all <-  sem(model_all, data = full, estimator = "MLR", se = "robust.huber.white")
s <- summary(fit_all, standardized = TRUE, fit.measures = TRUE, rsquare = TRUE)
p <- parameterEstimates(fit_all)
```

############################################################
# RDA
```{r}

rda.distichus.1 <- rda(dat3@tab ~., data = all_data2[, 22:ncol(all_data2)], scale = TRUE)
    RsquareAdj(rda.distichus.1)
    summary(rda.distichus.1)$concont

vif.cca(rda.distichus.1)

signif.rda.1 <- anova.cca(rda.distichus.1)

levels(all_data2$cluster) <- c("ignigularis", "east dominicensis", "South paleo-island", "properus", "west dominicensis", "ravitergum")
clust <- all_data2$cluster
bg <- c("#cf5d34", "#634102", "#993cc8", "gray80", "#c6b89c", "#fddc1f")

pdf("rda/rda-distichus-subsp-full-plot.pdf")
    plot(rda.distichus.1, type = "n", scaling = 3)
    points(rda.distichus.1, display = "species", pch = 20, cex = 0.7, col = "gray32", scaling = 3, choices = c(1, 2)) # SNPs in this case
    points(rda.distichus.1, display = "sites", pch = 21, cex = 1.3, col = "gray32", scaling = 3, bg = bg[as.factor(clust)], choices = c(1, 2)) # lizards
    text(rda.distichus.1, scaling = 3, display = "bp", col = "#0868ac", cex = 1) # predictor variables
    legend("topright", legend = levels(clust), bty = "n", col = "gray32", pch = 21, cex = 1, pt.bg = bg)
dev.off()
```